### PR TITLE
New Resource: `azurerm_monitor_scheduled_query_rules_alert_v2`

### PR DESCRIPTION
**Pull request by teowa**
Resolves https://github.com/hashicorp/terraform-provider-azurerm/issues/14999 , resolves https://github.com/hashicorp/terraform-provider-azurerm/issues/16219, resolves https://github.com/hashicorp/terraform-provider-azurerm/issues/17510, resolves https://github.com/hashicorp/terraform-provider-azurerm/issues/17544

Existing resources: [`azurerm_monitor_scheduled_query_rules_alert`](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/monitor_scheduled_query_rules_alert) and [`azurerm_monitor_scheduled_query_rules_log`](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/monitor_scheduled_query_rules_log), based on REST API [2018-04-16](https://github.com/Azure/azure-rest-api-specs/blob/main/specification/monitor/resource-manager/Microsoft.Insights/stable/2018-04-16/scheduledQueryRule_API.json)

Reference:
1. [Log Alert Introduction](https://docs.microsoft.com/en-us/azure/azure-monitor/alerts/alerts-types#log-alerts) and  [Log Alert Guides](https://docs.microsoft.com/en-us/azure/azure-monitor/alerts/alerts-log)
2. [REST API specification (2021-08-01)](https://github.com/Azure/azure-rest-api-specs/blob/main/specification/monitor/resource-manager/Microsoft.Insights/stable/2021-08-01/scheduledQueryRule_API.json)
3. [Why `v2`](https://docs.microsoft.com/en-us/azure/azure-monitor/alerts/alerts-common-schema-definitions#monitoringservice--log-alerts-v2)

Test:
```
make acctests SERVICE='monitor' TESTARGS='-run=TestAccMonitorScheduledQueryRulesAlertV2_' TESTTIMEOUT='60m'
==> Checking that code complies with gofmt requirements...
==> Checking that Custom Timeouts are used...
==> Checking that acceptance test packages are used...
TF_ACC=1 go test -v ./internal/services/monitor -run=TestAccMonitorScheduledQueryRulesAlertV2_ -timeout 60m -ldflags="-X=github.com/hashicorp/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccMonitorScheduledQueryRulesAlertV2_basic
=== PAUSE TestAccMonitorScheduledQueryRulesAlertV2_basic
=== RUN   TestAccMonitorScheduledQueryRulesAlertV2_requiresImport
=== PAUSE TestAccMonitorScheduledQueryRulesAlertV2_requiresImport
=== RUN   TestAccMonitorScheduledQueryRulesAlertV2_complete
=== PAUSE TestAccMonitorScheduledQueryRulesAlertV2_complete
=== RUN   TestAccMonitorScheduledQueryRulesAlertV2_update
=== PAUSE TestAccMonitorScheduledQueryRulesAlertV2_update
=== CONT  TestAccMonitorScheduledQueryRulesAlertV2_basic
=== CONT  TestAccMonitorScheduledQueryRulesAlertV2_complete
=== CONT  TestAccMonitorScheduledQueryRulesAlertV2_requiresImport
=== CONT  TestAccMonitorScheduledQueryRulesAlertV2_update
--- PASS: TestAccMonitorScheduledQueryRulesAlertV2_basic (170.58s)
--- PASS: TestAccMonitorScheduledQueryRulesAlertV2_complete (180.08s)
--- PASS: TestAccMonitorScheduledQueryRulesAlertV2_requiresImport (191.58s)
--- PASS: TestAccMonitorScheduledQueryRulesAlertV2_update (370.86s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/monitor       372.001s

```

### CODE REVIEW

File: website/docs/r/monitor_scheduled_query_rules_alert_v2.html.markdown [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+
+## Attributes Reference
+
+The following Attributes are exported:
+
+* `id` - The id of the Monitor Scheduled Query Rule.
+
+* `created_with_api_version` - The api-version used when creating this alert rule.
+
+* `is_legacy_log_analytics_rule` - True if this alert rule is a legacy Log Analytic Rule.
```
Comment (katbyte) [inline website/docs/r/monitor_scheduled_query_rules_alert_v2.html.markdown:172]:
```suggestion
* `is_a_legacy_log_analytics_rule` - True if this alert rule is a legacy Log Analytic Rule.
```

File: website/docs/r/monitor_scheduled_query_rules_alert_v2.html.markdown [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+
+* `operator` - (Required) Operator for dimension values. Possible values are `Exclude`,and `Include`.
+
+* `values` - (Required) List of dimension values. Use a wildcard `*` to collect all.
+
+---
+
+A `failing_periods` block supports the following:
+
+* `min_failing_periods_to_alert` - (Required) Specifies the number of violations to trigger an alert. Should be smaller or equal to `number_of_evaluation_periods`. Possible value is integer between 1 and 6.
```
Comment (katbyte) [inline website/docs/r/monitor_scheduled_query_rules_alert_v2.html.markdown:160]:
```suggestion
* `minimum_failing_periods_to_trigger_alert` - (Required) Specifies the number of violations to trigger an alert. Should be smaller or equal to `number_of_evaluation_periods`. Possible value is integer between 1 and 6.
```

File: website/docs/r/monitor_scheduled_query_rules_alert_v2.html.markdown [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+
+A `criteria` block supports the following:
+
+* `operator` - (Required) Specifies the criteria operator. Possible values are `Equals`, `GreaterThan`, `GreaterThanOrEqual`, `LessThan`,and `LessThanOrEqual`. 
+
+* `query` - (Required) The query to run on logs. The results returned by this query are used to populate the alert. 
+
+* `threshold` - (Required) Specifies the criteria threshold value that activates the alert.
+
+* `time_aggregation` - (Required) The type of aggregation to apply to the data points in aggregation granularity. Possible values are `Average`, `Count`, `Maximum`, `Minimum`,and `Total`.
```
Comment (katbyte) [inline website/docs/r/monitor_scheduled_query_rules_alert_v2.html.markdown:136]:
```suggestion
* `time_aggregation_methold` - (Required) The type of aggregation to apply to the data points in aggregation granularity. Possible values are `Average`, `Count`, `Maximum`, `Minimum`,and `Total`.
```

File: website/docs/r/monitor_scheduled_query_rules_alert_v2.html.markdown [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+
+* `criteria` - (Required) A `criteria` block as defined below.
+
+* `evaluation_frequency` - (Required) How often the scheduled query rule is evaluated, represented in ISO 8601 duration format. 
+
+* `scopes` - (Required) Specifies the list of resource ids that this scheduled query rule is scoped to. Changing this forces a new resource to be created.
+
+* `severity` - (Required) Severity of the alert. Should be an integer between 0 and 4. Value of 0 is severest. 
+
+* `window_size` - (Required) Specifies the period of time in ISO 8601 duration format on which the Scheduled Query Rule will be executed (bin size).
```
Comment (katbyte) [inline website/docs/r/monitor_scheduled_query_rules_alert_v2.html.markdown:92]: size is vauge
```suggestion
* `window_duration` - (Required) Specifies the period of time in ISO 8601 duration format on which the Scheduled Query Rule will be executed (bin size).
```

File: website/docs/r/monitor_scheduled_query_rules_alert_v2.html.markdown [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+
+* `action` - (Optional) An `action` block as defined below.
+
+* `auto_mitigation_enabled` - (Optional) Specifies the flag that indicates whether the alert should be automatically resolved or not. Value should be true or false. The default is false.
+
+* `check_workspace_alerts_storage_configured` - (Optional) Specifies the flag which indicates whether this scheduled query rule should be stored in the customer's storage. Value should be true or false. The default is false.
```
Comment (katbyte) [inline website/docs/r/monitor_scheduled_query_rules_alert_v2.html.markdown:98]: the property name indicates it checks something but the description states it does something?
```suggestion
* `check_if workspace_alerts_storage_is_configured` - (Optional) Specifies the flag which indicates whether this scheduled query rule check if storage is configured. Value should be true or false. The default is false. 
```
or
```suggestion
* `workspace_alerts_storage_enabled` - (Optional) Specifies the flag which indicates whether this scheduled query rule check if storage is configured. Value should be true or false. The default is false. 
```

File: website/docs/r/monitor_scheduled_query_rules_alert_v2.html.markdown [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+
+* `scopes` - (Required) Specifies the list of resource ids that this scheduled query rule is scoped to. Changing this forces a new resource to be created.
+
+* `severity` - (Required) Severity of the alert. Should be an integer between 0 and 4. Value of 0 is severest. 
+
+* `window_size` - (Required) Specifies the period of time in ISO 8601 duration format on which the Scheduled Query Rule will be executed (bin size).
+
+* `action` - (Optional) An `action` block as defined below.
+
+* `auto_mitigation_enabled` - (Optional) Specifies the flag that indicates whether the alert should be automatically resolved or not. Value should be true or false. The default is false.
```
Comment (katbyte) [inline website/docs/r/monitor_scheduled_query_rules_alert_v2.html.markdown:96]: all true false values should be quotes
```suggestion
* `auto_mitigation_enabled` - (Optional) Specifies the flag that indicates whether the alert should be automatically resolved or not. Value should be `true` or `false`. The default is `false`.
```

File: website/docs/r/monitor_scheduled_query_rules_alert_v2.html.markdown [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+
+* `check_workspace_alerts_storage_configured` - (Optional) Specifies the flag which indicates whether this scheduled query rule should be stored in the customer's storage. Value should be true or false. The default is false. 
+
+* `description` - (Optional) Specifies the description of the scheduled query rule.
+
+* `display_name` - (Optional) Specifies the display name of the alert rule.
+
+* `enabled` - (Optional) Specifies the flag which indicates whether this scheduled query rule is enabled. Value should be true or false. The default is true.
+
+* `mute_actions_duration` - (Optional) Mute actions for the chosen period of time in ISO 8601 duration format after the alert is fired.
```
Comment (katbyte) [inline website/docs/r/monitor_scheduled_query_rules_alert_v2.html.markdown:106]:
```suggestion
* `mute_actions_after_alert_duration` - (Optional) Mute actions for the chosen period of time in ISO 8601 duration format after the alert is fired. 
```

File: website/docs/r/monitor_scheduled_query_rules_alert_v2.html.markdown [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+
+* `display_name` - (Optional) Specifies the display name of the alert rule.
+
+* `enabled` - (Optional) Specifies the flag which indicates whether this scheduled query rule is enabled. Value should be true or false. The default is true.
+
+* `mute_actions_duration` - (Optional) Mute actions for the chosen period of time in ISO 8601 duration format after the alert is fired. 
+
+-> **NOTE** `auto_mitigation_enabled` and `mute_actions_duration` are mutually exclusive and cannot both be set.
+
+* `override_query_time_range` - (Optional) If specified then overrides the query time range, default is `window_size`*`number_of_evaluation_periods`.
```
Comment (katbyte) [inline website/docs/r/monitor_scheduled_query_rules_alert_v2.html.markdown:110]: starting with override implies a bool
```suggestion
* `query_time_range_override` - (Optional) If specified then overrides the query time range, default is `window_size`*`number_of_evaluation_periods`.
```

File: internal/services/monitor/monitor_scheduled_query_rules_alert_v2_resource.go [hunk 9]
```diff
@@ -0,0 +1,895 @@
...
+						ValidateFunc: validation.StringInSlice([]string{
+							string(scheduledqueryrules.TimeAggregationCount),
+							string(scheduledqueryrules.TimeAggregationAverage),
+							string(scheduledqueryrules.TimeAggregationMinimum),
+							string(scheduledqueryrules.TimeAggregationMaximum),
+							string(scheduledqueryrules.TimeAggregationTotal),
+						}, false),
+					},
+
+					"threshold": {
```
Comment (katbyte) [inline internal/services/monitor/monitor_scheduled_query_rules_alert_v2_resource.go:137]: is this a persntage?
```suggestion
					"threshold_percentage": {
```

**Review by katbyte — CHANGES_REQUESTED** [review]

File: website/docs/r/monitor_scheduled_query_rules_alert_v2.html.markdown (hunk 5 above)
Comment (teowa) [reply website/docs/r/monitor_scheduled_query_rules_alert_v2.html.markdown:98]: I think `workspace_alerts_storage_enabled` is better : )

File: website/docs/r/monitor_scheduled_query_rules_alert_v2.html.markdown (hunk 3 above)
Comment (teowa) [reply website/docs/r/monitor_scheduled_query_rules_alert_v2.html.markdown:136]: it should be `time_aggregation_method`

File: internal/services/monitor/monitor_scheduled_query_rules_alert_v2_resource.go (hunk 9 above)
Comment (teowa) [reply internal/services/monitor/monitor_scheduled_query_rules_alert_v2_resource.go:137]: This threshold can accpect float values.

**Comment by teowa** [comment]
Hi @katbyte , thanks for review. Per your comment, I have made changes or added a reply, please kindly take another look.

File: internal/services/monitor/monitor_scheduled_query_rules_alert_v2_resource.go (hunk 9 above)
Comment (katbyte) [reply internal/services/monitor/monitor_scheduled_query_rules_alert_v2_resource.go:137]: sure, could we add some validation thou?

Comment (katbyte) [reply internal/services/monitor/monitor_scheduled_query_rules_alert_v2_resource.go:137]: Makes sense to me!

**Review by katbyte — APPROVED** [review]
LGTM 🍁

**Pull request merged**